### ISSUE TITLE
[Feature request] Generate terraform schema for other languages

### ISSUE THREAD

**Issue by pie6k** [hashicorp/terraform#19144]
Imagine you can write tf infrastructure definition using other languages like TypeScript:

It would allow user to do things like:

```ts
// some reusable function that creates some resource in type-safe way
function createBigTableInstance(name: string, nodesCount: 3) {
  // note that in this case function interface would be the same as https://www.terraform.io/docs/providers/google/r/bigtable_instance.html
  // you would see errors in your config instantly, even before you save the file
  return new terraformSDK.gcp.bigTableInstance(name, {
    name: 'tf-instance',
    cluster: {
      clusterId: 'tf-instance-cluster',
      zone: 'us-central1-b',
      numNodes: nodesCount,
      storageType: 'HDD',
    },
  });
}

// other function that can reuse such function to create multiple resources
// as it's just nodejs javascript - imagination and js engine are the limits about what can be done
function createInfrastructure() {
  return [
    createBigTableInstance('a', 3),
    createBigTableInstance('b', 3),
    createBigTableInstance('c', 3),
  ]
}

// in the end - we can generate terraform compatible file to use for development, CI or CD
function generateInfrastructureFile() {
  const infrastructure = createInfrastructure();
  terraformSDK.saveConfig(infrastructure, 'infrastructure.tf');
}

generateInfrastructureFile();
```

However - to make such SDK usable in long-term - it needs to be in-sync with current terraform resources schema. Knowing how big this schema is - I think generating such interface in other language would have to be automated.

So final question is:

Is there any 'definition' of entire terraform resources schema that would define all possible resources with their data interface in a 'neutral' format that could be interpreted by other languages?

I believe being able to use 'real language' that would be in sync with terraform schema would open whole new way of defining infrastructure. 

Instead of rewriting complicated files for each service, you could create high-level, reusable, composable functions like `generateServiceWithDatabase('dogs-service').saveToFile('service.tf')`.

It could be next-level of 'infrastructure as a code'

----------------------

I've found this piece of great code by Pulumi - https://github.com/pulumi/pulumi-terraform/blob/master/pkg/tfgen/generate_nodejs.go - that seems to do something similar, but it looks more like a hacky way of doing that

**Comment by pie6k** [comment]
Any update?

**Comment by apparentlymart** [comment]
Hi @pie6k! Sorry we didn't catch this earlier.

In [Terraform v0.12.0-beta1](https://www.hashicorp.com/blog/announcing-terraform-0-1-2-beta1) you can find an experimental new command [`terraform providers schema -json`](https://www.terraform.io/docs/commands/providers/schema.html) that will export the most important parts of the schema (the names and type information) in JSON format for any providers that are used by the current configuration. It's possible that there will be minor changes to this format before it's considered final since we want to hear about experiences using it, but in practice I expect that the overall structure of it won't change.

Alternatively, once there are releases for all of the providers you are interested in that support the new plugin protocol for v0.12.0 (which adds the mechanism through which Terraform reads this information) you could potentially write your own "client" to a particular plugin that calls its [`GetSchema`](https://github.com/hashicorp/terraform/blob/dfe83d9ad5dbcf1ff9bf70d36875a04efb9638b5/internal/tfplugin5/tfplugin5.proto#L118) method to obtain the schema information directly, though sadly we do not yet have full documentation of this protocol and probably will not for a while longer since currently the SDK serves as the "documentation".

The command I mentioned will be included in the forthcoming v0.12.0 final release, and development of it is already complete, so I'm going to close this out to reflect that.
